internal/replication/blockreplicator_3node_test.go

```diff
@@ -0,0 +1,313 @@
+package replication_test
+
+import (
+	"github.com/golang/protobuf/proto"
+	"github.com/stretchr/testify/assert"
+	"os"
+	"testing"
+	"time"
+
+	"github.com/IBM-Blockchain/bcdb-server/pkg/types"
+	"github.com/stretchr/testify/require"
+)
+
+// Scenario:
+// - Start 3 nodes together, wait for leader,
+// - Stop the leader, wait for new leader,
+// - Stop the leader, remaining node has no leader.
+func TestBlockReplicator_3Node_StartCloseStep(t *testing.T) {
+	env := createClusterEnv(t, "info", 3)
+	require.Equal(t, 3, len(env.nodes))
+	defer os.RemoveAll(env.testDir)
+
+	// start 3 at once
+	for _, node := range env.nodes {
+		err := node.Start()
+		require.NoError(t, err)
+	}
+
+	// wait for an agreed leader
+	sameLeaderCond := func() bool {
+		return env.AgreedLeaderIndex() >= 0
+	}
+	assert.Eventually(t, sameLeaderCond, 30*time.Second, 100*time.Millisecond)
+
+	//close the leader, wait for some node to become a new leader
+	leaderIndex1 := env.FindLeaderIndex()
+	require.True(t, leaderIndex1 >= 0)
+	t.Logf("Leader #1 index: %d", leaderIndex1)
+	err := env.nodes[leaderIndex1].Close()
+	require.NoError(t, err)
+	isLeaderCond := func() bool {
+		return env.FindLeaderIndex() >= 0
+	}
+	assert.Eventually(t, isLeaderCond, 30*time.Second, 100*time.Millisecond)
+
+	//close the 2nd leader, remaining node has no leader
+	leaderIndex2 := env.FindLeaderIndex()
+	require.True(t, leaderIndex2 >= 0)
+	t.Logf("Leader #2 index: %d", leaderIndex2)
+	err = env.nodes[leaderIndex2].Close()
+	require.NoError(t, err)
+
+	//remaining node has no leader
+	var lastIndex int
+	for i, _ := range env.nodes {
+		if i == leaderIndex1 || i == leaderIndex2 {
+			continue
+		}
+		lastIndex = i
+	}
+	noLeaderCond := func() bool {
+		err := env.nodes[lastIndex].blockReplicator.IsLeader()
+		if err != nil && err.LeaderID == 0 {
+			return true
+		}
+		return false
```

Comment:
If we use `IsLeader() bool`, we can do
```suggestion
		return !env.nodes[lastIndex].blockReplicator.IsLeader()
```

Reply:
See above, `IsLeader() error` is needed by redirection. Moreover, it is not the same - here I check that lastIndex is not the leader AND that it does not know who the leader is; i.e. it is in a minority partition.